falcon/util/uri.py

```diff
@@ -292,9 +292,9 @@ def parse_query_string(query_string, keep_blank_qs_values=False):
             # a list, or append the next value to the list.
             old_value = params[k]
             if isinstance(old_value, list):
-                old_value.append(v)
+                old_value.append(decode(v))
             else:
-                params[k] = [old_value, v]
+                params[k] = [old_value, decode(v)]
 
         else:
             if ',' in v:
```

Comment:
In fact, making this factoring will don't fix original problem.

The only thing to determine if a comma is the reserved character or a simple character in your value is the percent-encoding. So, if i want to make a comma-separated list, i use the reserved character ','. Otherwise, if i want to send a value with a comma (like json, complex url, etc ...), i use  percent-encoding.
For me, falcon has to decode value at the last moment. So query parser will not handle all percent-encoded comma. Like that it won't confound with this sub-delimiter.

ex:
I want send the json '{"test1": "data1", "test2": "data2"}' by query parameter.
So we encode it : %7B%22test1%22%3A%20%22data1%22%
if we decode value before checking commas, the result will be a list: ['{"test1": "data1"', ' "test2": "data2"}'] while we wait a string.

As i understand URI rfc:
http://tools.ietf.org/html/rfc3986#section-2.4
"When a URI is dereferenced, the components and subcomponents significant to the scheme-specific dereferencing process (if any) must be parsed and separated before the percent-encoded octets within those components can be safely decoded, as otherwise the data may be mistaken for component delimiters"
http://tools.ietf.org/html/rfc3986#section-7.3
"Applications must split the URI into its components and subcomponents prior to decoding the octets, as otherwise the decoded octets might be mistaken for delimiters."